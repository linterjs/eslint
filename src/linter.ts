import {
  FormatInput,
  FormatOutput,
  LinterAdapter,
  LinterFactory,
  LinterProvider,
  LintInput,
  LintOutput,
} from "@linter/core";
import { CLIEngine } from "eslint";
import { getLogLevel, logger } from "./logger";
import { eslintReportToFormatOutput, eslintReportToLintOutput } from "./utils";

class Linter implements LinterAdapter {
  private cliEngine = new CLIEngine({
    fix: () => this.fix,
  });
  private fix = false;

  constructor() {
    logger.debug("Initializing Linter");

    const logLevel = getLogLevel();
    logger.debug(`Setting log level to "${logLevel}"`);
    logger.setLevel(logLevel);
  }

  public format({ filePath, text }: FormatInput): FormatOutput {
    logger.debug("Running format");
    filePath && logger.debug(`filePath: ${filePath}`);
    logger.debug(`text: ${text}`);
    logger.debug('Setting fix to "true"');
    this.fix = true;

    logger.debug("Running ESLint");
    const report = this.cliEngine.executeOnText(text, filePath);

    logger.debug('Setting fix to "false"');
    this.fix = false;

    const formatOutput = eslintReportToFormatOutput(report);
    logger.debug("Format done");
    return formatOutput;
  }

  public lint({ filePath, text }: LintInput): LintOutput {
    logger.debug("Running lint");
    filePath && logger.debug(`filePath: ${filePath}`);
    logger.debug(`text: ${text}`);

    logger.debug("Running ESLint");
    const report = this.cliEngine.executeOnText(text, filePath);

    const lintOutput = eslintReportToLintOutput(report);
    logger.debug("Format done");
    return lintOutput;
  }
}

const linterFactory: LinterFactory = () => new Linter();

const linterProvider: LinterProvider = {
  factory: linterFactory,
  name: "eslint",
  supportedExtensions: ["js", "jsx"],
};

export { Linter, linterFactory, linterProvider };
