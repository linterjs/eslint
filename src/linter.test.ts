import fs from "jest-plugin-fs";
import { Linter, linterProvider, linterFactory } from "./linter";

jest.mock("fs", () => {
  const realFs = require.requireActual("fs");
  const mockFs = require("jest-plugin-fs/mock");
  const { ufs } = require("unionfs");
  ufs.use(realFs).use(mockFs);
  return ufs;
});

beforeAll(() => {
  fs.root = process.cwd();
});

afterEach(() => {
  fs.restore();
  jest.resetModules();
});

test("Create linter", async () => {
  const linter = await linterProvider.linterFactory();
  expect(linter).toBeInstanceOf(Linter);
});

describe("Format", () => {
  test("with text", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
    });

    const linter = await linterProvider.linterFactory();

    expect(linter.format({ text: 'var foo = "bar"' })).toMatchSnapshot();
  });

  test("with text and filePath", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
    });

    const linter = await linterProvider.linterFactory();

    expect(
      linter.format({ filePath: "./foo.js", text: 'var bar = "baz"' }),
    ).toEqual({
      errorCount: 0,
      filePath: `${process.cwd()}/foo.js`,
      messages: [],
      output: "var bar = 'baz';",
      warningCount: 0,
    });
  });

  test("with already formatted text", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
    });

    const linter = await linterProvider.linterFactory();

    expect(linter.format({ text: "var foo = 'bar';" })).toMatchSnapshot();
  });

  test("with ignored file", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
      ".eslintignore": "foo.js",
    });

    const linter = await linterProvider.linterFactory();

    expect(
      linter.format({ filePath: "foo.js", text: 'var bar = "baz"' }),
    ).toMatchSnapshot();
  });
});

describe("Lint", () => {
  test("with text", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
    });

    const linter = await linterProvider.linterFactory();

    expect(linter.lint({ text: 'var foo = "bar"' })).toMatchSnapshot();
  });

  test("with text and filePath", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
    });

    const linter = await linterProvider.linterFactory();

    expect(
      linter.lint({ filePath: "foo.js", text: 'var bar = "baz"' }),
    ).toEqual({
      errorCount: 1,
      filePath: `${process.cwd()}/foo.js`,
      messages: [
        {
          column: 11,
          endColumn: 16,
          endLine: 1,
          line: 1,
          message: "Strings must use singlequote.",
          ruleId: "quotes",
          severity: 2,
        },
        {
          column: 16,
          line: 1,
          message: "Missing semicolon.",
          ruleId: "semi",
          severity: 1,
        },
      ],
      warningCount: 1,
    });
  });

  test("with syntax error", async () => {
    fs.mock({
      ".eslintrc": JSON.stringify({
        rules: {
          quotes: ["error", "single"],
          semi: ["warn"],
        },
      }),
    });

    const linter = await linterProvider.linterFactory();

    expect(linter.lint({ text: 'var foo ==== "bar"' })).toMatchSnapshot();
  });
});
