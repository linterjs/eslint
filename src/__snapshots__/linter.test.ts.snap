// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Format with already formatted text 1`] = `
Object {
  "errorCount": 0,
  "filePath": "<text>",
  "messages": Array [],
  "warningCount": 0,
}
`;

exports[`Format with ignored file 1`] = `
Object {
  "errorCount": 0,
  "messages": Array [],
  "warningCount": 0,
}
`;

exports[`Format with text 1`] = `
Object {
  "errorCount": 0,
  "filePath": "<text>",
  "messages": Array [],
  "output": "var foo = 'bar';",
  "warningCount": 0,
}
`;

exports[`Lint with syntax error 1`] = `
Object {
  "errorCount": 1,
  "filePath": "<text>",
  "messages": Array [
    Object {
      "column": 9,
      "line": 1,
      "message": "Parsing error: Unexpected token ===",
      "ruleId": "parser",
      "severity": 2,
    },
  ],
  "warningCount": 0,
}
`;

exports[`Lint with text 1`] = `
Object {
  "errorCount": 1,
  "filePath": "<text>",
  "messages": Array [
    Object {
      "column": 11,
      "endColumn": 16,
      "endLine": 1,
      "line": 1,
      "message": "Strings must use singlequote.",
      "ruleId": "quotes",
      "severity": 2,
    },
    Object {
      "column": 16,
      "line": 1,
      "message": "Missing semicolon.",
      "ruleId": "semi",
      "severity": 1,
    },
  ],
  "warningCount": 1,
}
`;
